contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -4235,6 +4239,64 @@ pg_proc_update_oid_acl(ObjectAddress address, Oid oid, Acl *acl)
 	table_close(rel, RowExclusiveLock);
 }
 
+/*
+ * Update the function definition of an alter procedure query 
+ * "ALTER ..." to "CREATE ..." in bbf_function_ext
+ */
+static void bbf_func_ext_update_proc_definition(Oid oid)
+{
+	Relation	bbf_function_ext_rel;
+	TupleDesc	bbf_function_ext_rel_dsc;
+	Datum 		new_record[BBF_FUNCTION_EXT_NUM_COLS];
+	bool		new_record_replaces[BBF_FUNCTION_EXT_NUM_COLS];
+	bool		new_record_nulls[BBF_FUNCTION_EXT_NUM_COLS];
+	char 		*original_query = get_original_query_string();
+	HeapTuple	tuple,
+				proctup,
+				oldtup;
+	
+	
+	proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(oid));
+	if (!HeapTupleIsValid(proctup)) // throw error
+		return;
+	oldtup = get_bbf_function_tuple_from_proctuple(proctup);
+	bbf_function_ext_rel = table_open(get_bbf_function_ext_oid(), RowExclusiveLock);
+	bbf_function_ext_rel_dsc = RelationGetDescr(bbf_function_ext_rel);
+
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+	MemSet(new_record_replaces, false, sizeof(new_record_replaces));
+
+	if (HeapTupleIsValid(oldtup))
+	{
+		StringInfoData infoSchemaStr;
+		initStringInfo(&infoSchemaStr);
+
+		/*
+		 * This solution only works because original_query does not contain
+		 * any leading characters or comments before "ALTER". When BABEL-5140
+		 * is resolved we will need to refactor this code
+		 */
+		if(strlen(original_query) >= 5 && strncasecmp(original_query, "alter", 5) == 0)
+		{
+			/* Change alter to create, add rest of characters, and update */
+			appendStringInfoString(&infoSchemaStr, "CREATE");
+			appendStringInfoString(&infoSchemaStr, original_query + 5);
+			new_record[Anum_bbf_function_ext_definition - 1] = CStringGetTextDatum(infoSchemaStr.data);
+			new_record_replaces[Anum_bbf_function_ext_definition - 1] = true;
+		}
+
+		tuple = heap_modify_tuple(oldtup, bbf_function_ext_rel_dsc,
+								  new_record, new_record_nulls,
+								  new_record_replaces);
```

Comment:
If the above `if` statement is false, we may call `heap_modify_tuple()` with a garbage `new_record` value. Is this an expected behavior? Do we have a such test case?

Reply:
new_record_replaces is initialized to false and only is true when the if statement is hit. Therefore if this function were run with an original query that does not start with "alter" nothing would update. This case should never happen though because bbf_func_ext_update_proc_definition is called from the alter procedure case

Reply:
If you think this case should never happen then please change the if condition to `Assert()`. In that way, you give a clear context and assumption for future developers.